### ISSUE TITLE
if.bind in some cases causing the bound variable to lose context???

### ISSUE THREAD

**Issue by klabranche** [aurelia/templating-resources#271]
The issue title is suspect.  I'm not sure "what is going on"....

**I'm submitting a bug report**

* **Library Version:**
latest release of as 2/2/2017 for all....  used au new to create project
1.0.8 (framework), 1.2.0 (templating-resources)

**Please tell us about your environment:**
* **Operating System:**
Windows 7

* **Node Version:**
6.9.4

* **NPM Version:**
3.5.2

* **JSPM OR Webpack AND Version**
JSPM 

* **Browser:**
all presumably but tested on IE 11, Chrome 55, Firefox 50 

* **Language:**
all presumably but only tested with TypeScript 2.1.5

**Current behavior:**
https://gist.run/?id=a54ec9b9638d86bf5794119c75217c49 
In the gist, I have an app class with a selectedHero.  The app.html has a list of heroes and an if binding to control the viewing of the selected hero. 

Depending on two factors I get different results with the list's class test to set the selected style to highlight the currently selected hero in the list.

Factor 1 - How I instantiate the selectedHero variable.
If I instantiate using  "selectedHero: Hero;" in the app class the if binding "appears"  to do something to the selectedHero variable to break the list's class test.  I have found in this context the test  class="${selectedHero === hero ? 'selected':''}" always returns false.

To correct, I instantiate the selectedHero to null "selectedHero: Hero = null;" and the highlighting works.

Factor 2 - Timing appears to be a factor.

Introducing a delay in the setting of the heroes array in the App class appears to influence this issue as well.  In the original gist (Factor 1) I am setting the heroes variable without any delay.  

If I add a delay to the setting of the heroes array then the code works as expected (the test to set the highlight css works) with or without the instantiation differences of the selectedHero variable. 

https://gist.run/?id=b913c1a449a05c7635904b33eb97a990

In the above gist I have added a 1 second delay in setting the heroes array of the App class.  I have kept the selectedHero instantiation to not setting it to null, rather just "selectedHero: Hero;"

If I remove the delay the test to set the css class to highlight the selectedHero from the list fails.

**Expected/desired behavior:**
I would expect the highlight test to work w/wo the dependence on how I instantiate my variable and/or w/wo a delay in the setting of the heroes array.

* **What is the motivation / use case for changing the behavior?**

I have no idea if this is a bug or if I am misunderstanding something about the usage of "if" and/or javascript itself.  

I posed this question on the Aurelia gitter channel and a few thought this behavior was "interesting" enough wo a clear reason as to why.

Ashley Grant and another thought it was important enough to warrant creating this issue.  So here we are.... :-)

I have a repo of my tour of heroes project where I found this on if you wish to look at it: https://github.com/klabranche/TourOfHeroesAurelia

I have also verified that this issue is present in Ashley Grant's tour of heroes example: https://github.com/AshleyGrant/tour-of-heroes

Thanks for all you do, keep up the great work.  Aurelia is awesome & I wish you all the best.

Thanks,
Kevin

**Comment by jdanyow** [comment]
*tldr: it's important to make sure you define properties you're binding to when your bindings are not explicit about which scope the property is defined on (eg `selectedHero` vs `$parent.selectedHero`)*

Consider the JavaScript statement `foo.bar()`. To execute that statement the V8/Chakra/etc needs to find `foo` by checking whether it's defined in the current scope or an outer scope, before returning the value of foo or throwing *foo is not defined* if the variable is not defined.

The binding engine does something similar when searching for the `selectedHero` property. 

Using your gist as an example, and removing the `setTimeout` logic so the `heroes` property is immediately defined, consider this part of the repeat template: `class="${selectedHero === hero ? 'selected':''}"`. When evaluating that binding Aurelia will first try to find `selectedHero` on the repeat item scope. All it will find are the following properties:
* hero
* $first
* $last
* $index
* $odd
* $even

Since Aurelia didn't find `selectedHero` on the repeat item scope it will search the outer scope which is the App view model instance. All it will find there is `title` and `heroes` because `selectedHero` has not yet been defined.
* title
* heroes
* ~~selectedHero~~ (not yet defined)

Again, `selectedHero` wasn't found, and there's not another outer scope to search. Unlike JavaScript, the binding system doesn't throw an error when a property is not defined. Instead, the binding system assumes `selectedHero` will eventually become defined on the closest scope and starts observing that property on the closest scope. In this case, the closest scope is the repeat item scope. In reality `selectedHero` is going to become defined on the outer app instance scope. Later on, when a hero is selected by the user and `App.selectedHero` becomes defined, nothing will happen because the binding is observing `selectedHero` on the repeat item scope, not the App scope.

---

Now let's consider the case where `setTimeout` is used to delay setting the `heroes` array. In this case, the array the repeat is bound to is undefined so the `class="${selectedHero === hero ? 'selected':''}"` binding in the repeat template is not evaluated. Skipping down to the `<div if.bind="selectedHero">` part of the template, Aurelia searches the closest scope (the app instance) for `selectedHero` and doesn't find it. Just as before the binding system assumes `selectedHero` will eventually become defined on the closest scope and starts observing that property on the closest scope (an instance of the App class). Here's the twist... observing a property has side-effects. Aurelia uses a strategy we call "setter observation" to detect when a property changes. When a property is observed, Aurelia uses `Object.defineProperty` to rewrite the property with a getter and setter method. The setter method allows us to detect when the property is assigned and update any associated bindings. *Observing `selectedHero` on the App viewmodel instance causes the binding system to define the property!* Later, when the `setTimeout` calls back and `heroes` becomes defined, the repeat template will be evaluated, and this time `selectedHero` is defined so the binding system won't get into the situation from the previous example where it had to make an assumption where `selectedHero` is going to become defined.

Let me know if you have any question on this stuff.

cc @AshleyGrant

**Comment by AshleyGrant** [comment]
The thing I'm still not understanding is what causes the difference in behavior between `show` and `if`?

**Comment by klabranche** [comment]
@jdanyow  Wow!  I just learned me a bunch!  Thanks!!!!  I also wonder about `show` apparently appearing to be unaffected by my examples.

**Comment by jdanyow** [comment]
'if' and 'repeat' are template controllers, when the view is rendered they are evaluated after the normal bindings and custom attributes like 'show'. Using 'show' would have caused selectedHero to become defined before the repeat rendered.

https://github.com/aurelia/templating/blob/master/src/view.js#L144

**Comment by klabranche** [comment]
Awesome!  Thanks so much for the clarity. :-)

**Comment by CasiOo** [comment]
Thanks so much @jdanyow . Great explanation!
